## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
#columns_to_drop = ['EIN', 'NAME']
c_t_drop = application_df.drop(['EIN'], axis=1)
c_t_drop.head()


NAME APPLICATION_TYPE  \
0              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
# Determine the number of unique values in each column.
c_t_drop.nunique()

#or
# for column in c_t_drop.columns:
#     ctd_uniq_vals = c_t_drop[column].nunique()

#     print(column, ctd_uniq_vals)


NAME                      19568
APPLICATION_TYPE             17
AFFILIATION                   6
CLASSIFICATION               71
USE_CASE                      5
ORGANIZATION                  4
STATUS                        2
INCOME_AMT                    9
SPECIAL_CONSIDERATIONS        2
ASK_AMT                    8747
IS_SUCCESSFUL                 2
dtype: int64

In [4]:
# Look at NAME value counts for binning
name_cnts = c_t_drop['NAME'].value_counts()
name_cnts

PARENT BOOSTER USA INC                                                  1260
TOPS CLUB INC                                                            765
UNITED STATES BOWLING CONGRESS INC                                       700
WASHINGTON STATE UNIVERSITY                                              492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                          408
                                                                        ... 
ST LOUIS SLAM WOMENS FOOTBALL                                              1
AIESEC ALUMNI IBEROAMERICA CORP                                            1
WEALLBLEEDRED ORG INC                                                      1
AMERICAN SOCIETY FOR STANDARDS IN MEDIUMSHIP & PSYCHICAL INVESTIGATI       1
WATERHOUSE CHARITABLE TR                                                   1
Name: NAME, Length: 19568, dtype: int64

In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`

# Chose my cutoff value
cutoff_value = 5

names_to_replace = name_cnts[name_cnts <= cutoff_value].index.tolist()

# Replace in dataframe
for app in names_to_replace:
    c_t_drop['NAME'] = c_t_drop['NAME'].replace(app,"Other")

# Check to make sure binning was successful
c_t_drop['NAME'].value_counts()

Other                                           20043
PARENT BOOSTER USA INC                           1260
TOPS CLUB INC                                     765
UNITED STATES BOWLING CONGRESS INC                700
WASHINGTON STATE UNIVERSITY                       492
                                                ...  
HABITAT FOR HUMANITY INTERNATIONAL                  6
DAMAGE PREVENTION COUNCIL OF TEXAS                  6
FLEET RESERVE ASSOCIATION                           6
HUGH OBRIAN YOUTH LEADERSHIP                        6
INTERNATIONAL CONGRESS OF CHURCHES MINISTERS        6
Name: NAME, Length: 355, dtype: int64

In [6]:
# Look at APPLICATION_TYPE value counts for binning
app_type_cnts = c_t_drop['APPLICATION_TYPE'].value_counts()
app_type_cnts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [7]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`

# Chose my cutoff value
cutoff_value = 500

application_types_to_replace = app_type_cnts[app_type_cnts < cutoff_value].index.tolist()

# Replace in dataframe
for app in application_types_to_replace:
    c_t_drop['APPLICATION_TYPE'] = c_t_drop['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
c_t_drop['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [8]:
# Look at CLASSIFICATION value counts for binning
classification_vlu_cnts = c_t_drop.CLASSIFICATION.value_counts()
classification_vlu_cnts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [9]:
# You may find it helpful to look at CLASSIFICATION value counts >1
# Filter value counts greater than one
classification_vlu_cnts[classification_vlu_cnts > 1]


C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [10]:
classification_vlu_cnts[classification_vlu_cnts > 10]

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
Name: CLASSIFICATION, dtype: int64

In [11]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
#cutoff_value = 800

classifications_to_replace = classification_vlu_cnts[classification_vlu_cnts < 1000].index.tolist()

# Replace in dataframe
for cls in classifications_to_replace:
    c_t_drop['CLASSIFICATION'] = c_t_drop['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
c_t_drop['CLASSIFICATION'].value_counts()


C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [12]:
# Convert categorical data to numeric with `pd.get_dummies`
# Convert categorical data to numeric using one-hot encoding
df_cat_to_num = pd.get_dummies(c_t_drop)
df_cat_to_num

STATUS   ASK_AMT  IS_SUCCESSFUL  NAME_AACE INTERNATIONAL  \
0           1      5000              1                        0   
1           1    108590              1                        0   
2           1      5000              0                        0   
3           1      6692              1                        0   
4           1    142590              1                        0   
...       ...       ...            ...                      ...   
34294       1      5000              0                        0   
34295       1      5000              0                        0   
34296       1      5000              0                        0   
34297       1      5000              1                        0   
34298       1  36500179              0                        0   

       NAME_ACE MENTOR PROGRAM OF AMERICA INC  \
0                                           0   
1                                           0   
2                                           0   
3                                           0   
4                                           0   
...                                       ...   
34294                                       0   
34295                                       0   
34296                                       0   
34297                                       0   
34298                                       0   

       NAME_AFRICAN-AMERICAN POSTAL LEAGUE UNITED FOR SUCCESS A-PLUS  \
0                                                      0               
1                                                      0               
2                                                      0               
3                                                      0               
4                                                      0               
...                                                  ...               
34294                                                  0               
34295                                                  0               
34296                                                  0               
34297                                                  0               
34298                                                  0               

       NAME_AIR FORCE ASSOCIATION  NAME_ALABAMA FEDERATION OF WOMENS CLUBS  \
0                               0                                        0   
1                               0                                        0   
2                               0                                        0   
3                               0                                        0   
4                               0                                        0   
...                           ...                                      ...   
34294                           0                                        0   
34295                           0                                        0   
34296                           0                                        0   
34297                           0                                        0   
34298                           0                                        0   

       NAME_ALABAMA TREASURE FOREST ASSOCIATION  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   
...                                         ...   
34294                                         0   
34295                                         0   
34296                                         0   
34297                                         0   
34298                                         0   

       NAME_ALBANY STATE UNIVERSITY NATIONAL ALUMNI ASSOCIATION  ...  \
0                                                      0         ...   
1                                                     

In [20]:
#predict whether applicants will be successful if funded by Alphabet Soup.
# Split our preprocessed data into our features and target arrays. 21.Class2.Act6.getting_real_solution.ipynb
y = df_cat_to_num["IS_SUCCESSFUL"].values
X = df_cat_to_num.drop(["IS_SUCCESSFUL", "STATUS"], axis='columns').values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [21]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [22]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0])
hidden_nodes_layer1 =  90
hidden_nodes_layer2 = 30
hidden_nodes_layer3 = 10

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="sigmoid"))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="sigmoid"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_4 (Dense)             (None, 90)                35820     
                                                                 
 dense_5 (Dense)             (None, 30)                2730      
                                                                 
 dense_6 (Dense)             (None, 30)                930       
                                                                 
 dense_7 (Dense)             (None, 1)                 31        
                                                                 
Total params: 39,511
Trainable params: 39,511
Non-trainable params: 0
_________________________________________________________________


In [16]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [17]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=100)

Epoch 1/100
804/804 [==============================] - 4s 3ms/step - loss: 0.4889 - accuracy: 0.7633
Epoch 2/100
804/804 [==============================] - 2s 2ms/step - loss: 0.4384 - accuracy: 0.7921
Epoch 3/100
804/804 [==============================] - 2s 2ms/step - loss: 0.4317 - accuracy: 0.7942
Epoch 4/100
804/804 [==============================] - 2s 2ms/step - loss: 0.4280 - accuracy: 0.7971
Epoch 5/100
804/804 [==============================] - 2s 2ms/step - loss: 0.4262 - accuracy: 0.7971
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4248 - accuracy: 0.7971
Epoch 7/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4227 - accuracy: 0.7975
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.4220 - accuracy: 0.7992
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.4206 - accuracy: 0.7984
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.4201 - accura

In [18]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.4473 - accuracy: 0.7880 - 549ms/epoch - 2ms/step
Loss: 0.4473232328891754, Accuracy: 0.7879883646965027


In [19]:
# Export our model to HDF5 file
nn.save("AlphabetSoupCharity_Optimization.h5")

The purpose of this analysis was to predict whether applicants would be successful if funded by Alphabet Soup.

RESULTS
DATA PREPROCESSING
The target variables for my model included participating organizations' Names (PARENT BOOSTER USA INC-1260), Application Types (T3-27037), and Classification (C1000-17326) specifically those with the highest value counts.

The feature variables included AFFILIATION, USE_CASE, ORGANIZATION, STATUS, INCOME_AMT, SPECIAL_CONSIDERATIONS, and ASK_AMT. The variables I removed are IS_SUCCESSFUL and STATUS. because I want only for the model to make the prediction for applicants' successes if funded by Alphabet Soup.

COMPLILING, TRAINING, EVALUATING THE MODEL
The number of neurons selected in the first, second, and third layers were selected randomly to observe for a high accuracy prediction rate. The more neurons involved, the more the model's performance can be enhanced.
The Activation Function Sigmoid was mostly used because it was cited as being the most commonly used in neural networks due to its usefulness in predicting probabilities. I was able to achieve a fairly good targeted model performance with the use of three targets and three neuronal layers at 79%. To increase my model performances, I would use the same factors as above and maybe alternate with the columns that can be dropped from the dataset to see how the accuracy prediction value.

Overall: The model did pretty well. The output of an accuracy percentage of 79% predicting that specific applicant organizations with specific classifications and application types would be successful if funded by Alphabet Soup. Random Forest Model is recommended as a different model to use because it is another classification model but my current model was successful in giving me 79% accuracy prediction.